Improve cmake build for MSVC

sphaerophoria opened the pull request:
Current cmake files don't handle MSVC builds very well. They don't support toxav and only support libs being in a specific location. I've adjusted them to build on MSVC for me a lot better.

1. Setup a root prefix semi like linux would (include/ lib/ bin/). Copy pthread libs to lib and headers to include.
2. Download latest libsodium and extract it
3. build libvpx and libopus, installing them to the same directory we put pthreads in
3. `cmake .. -DCMAKE_INSTALL_PREFIX="<dir with pthreads/opus/vpx>" -DCMAKE_PREFIX_PATH="<path to libsodium extracted binaries>" -G"Visual Studio 15 2017 Win64"
4. cmake --build . --target install --config Release

After this I was able to build/link qtox and have an audio/video call with a friend on linux using the toxlib I built.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/toktok/c-toxcore/1191)
<!-- Reviewable:end -->

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/TokTok/c-toxcore?pullRequest=1191) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/TokTok/c-toxcore?pullRequest=1191) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/TokTok/c-toxcore?pullRequest=1191) it.</sub>

sphaerophoria commented:
As mentioned in the PR, the flow for setting up dependencies is a little different. As a result the CI build for MSVC failed. This is an expected result of the changeset.

iphydf commented:
@sphaerophoria can you update the appveyor instructions to set up the dependencies in the new way?

nurupo commented:
Since you have changed the way dependencies are being searched, I suggest you enable all of the Travis-CI builds to make sure all of them succeed. We have macOS build, FreeBSD build, Windows cross-compilation build, etc. It's quite possible that changing how dependencies are searched could have broken some of them. Just make a separate commit removing all of `if: type IN (push, api, cron)` lines in `.travis.yml`, push it, wait for Travis-CI to build, fix any failed builds, then once all the builds pass link the Travis build URL in a comment so that we could see them passing and remove that commit that modified `.travis.yml`.

sphaerophoria commented:
Thanks for the feedback, will do, wouldn't want to break other builds!

nurupo commented:
The Windows cross-compilation builds are failing to find libpthread.

The library is located at

```bash
$ find /usr/i686-w64-mingw32 -name '*pthread*' -name '*pthread*'
/usr/i686-w64-mingw32/include/pthread.h
/usr/i686-w64-mingw32/include/pthread_unistd.h
/usr/i686-w64-mingw32/include/pthread_signal.h
/usr/i686-w64-mingw32/include/pthread_time.h
/usr/i686-w64-mingw32/include/pthread_compat.h
/usr/i686-w64-mingw32/lib/libpthread.dll.a
/usr/i686-w64-mingw32/lib/libpthread.a
/usr/i686-w64-mingw32/lib/libwinpthread.dll.a
/usr/i686-w64-mingw32/lib/libwinpthread.a
/usr/i686-w64-mingw32/lib/libwinpthread-1.dll
$ find /usr/x86_64-w64-mingw32 -name '*pthread*' -name '*pthread*'
/usr/x86_64-w64-mingw32/include/pthread.h
/usr/x86_64-w64-mingw32/include/pthread_unistd.h
/usr/x86_64-w64-mingw32/include/pthread_signal.h
/usr/x86_64-w64-mingw32/include/pthread_time.h
/usr/x86_64-w64-mingw32/include/pthread_compat.h
/usr/x86_64-w64-mingw32/lib/libpthread.dll.a
/usr/x86_64-w64-mingw32/lib/libpthread.a
/usr/x86_64-w64-mingw32/lib/libwinpthread.dll.a
/usr/x86_64-w64-mingw32/lib/libwinpthread.a
/usr/x86_64-w64-mingw32/lib/libwinpthread-1.dll
```

The build scripts tell pkg-config and CMake's own find utilities to look into `/usr/i686-w64-mingw32` and `/usr/x86_64-w64-mingw32` by setting `PKG_CONFIG_PATH` env variable and `CMAKE_FIND_ROOT_PATH` CMake variable. Here is an example for i686:

```
export PKG_CONFIG_PATH=/root/prefix/i686/lib/pkgconfig:/root/extra-prefix/i686/lib/pkgconfig
echo "
        SET(CMAKE_SYSTEM_NAME Windows)

        SET(CMAKE_C_COMPILER   i686-w64-mingw32-gcc)
        SET(CMAKE_CXX_COMPILER i686-w64-mingw32-g++)
        SET(CMAKE_RC_COMPILER  i686-w64-mingw32-windres)

        SET(CMAKE_FIND_ROOT_PATH /usr/i686-w64-mingw32 /root/prefix/i686 /root/extra-prefix/i686)
" > windows_toolchain.cmake
cmake -DCMAKE_TOOLCHAIN_FILE=windows_toolchain.cmake [...]
```

nurupo commented:
The build scripts for this are in [`other/docker/windows`](https://github.com/TokTok/c-toxcore/tree/master/other/docker/windows). It's preferrable if the build scripts don't have to be modified, since it looks like we are properly telling pkg-config and CMake where to look for the libs.

sphaerophoria commented:
@nurupo I've replaced the Threads::Threads module with one that I thought was equivalent and also handles win32 pthreads correctly. It's quite possible that I accidentally overrode the mingw version as they're both win32 but are handled differently. 

I'll do my best to avoid touching the build scripts and adapting my changes to avoid changing behavior for non MSVC targets as much as possible.

sphaerophoria commented:
Alright CIs passed

https://travis-ci.org/TokTok/c-toxcore/builds/431320188?utm_source=github_status&utm_medium=notification
https://ci.appveyor.com/project/iphydf/c-toxcore/build/1.0.4567

I've squashed down the last commit and stripped the travis. Had to tell Findsodium.cmake that I wanted the static variant and only enable pthreads-win32 on msvc.

nurupo commented:
Would it still work if you have only shared sodium library? libtoxcore is packaged in Debian Sid and the libsodium is a shared library there.

sphaerophoria commented:
That's a good point, do you happen to know what pkg_use_module did under the hood to prioritize shared vs static linking before? I can change my PR to try one and fallback on the next

sphaerophoria commented:
I've updated the PR to try to link sodium statically before linking it shared.

sphaerophoria commented:
Closing/Reopening PR to re-trigger travis since it doesn't look like my change introduced the failure.

iphydf commented:
I don't think that works. You can push an amended commit, that triggers it.

iphydf commented:
I've restarted the build.

sphaerophoria commented:
Thanks, I think the PR is ready to go now.

nurupo commented:
>That's a good point, do you happen to know what pkg_use_module did under the hood to prioritize shared vs static linking before?

It used Cmake's PkgConfig module

https://github.com/TokTok/c-toxcore/blob/c8359c843b3f2a693e5c9ff47e1a9a0f288fa26d/cmake/ModulePackage.cmake#L15-L17

Here is documentation for it: https://cmake.org/cmake/help/v3.0/module/FindPkgConfig.html

It describesn a common case and a case when you explicitly ask for static library. We used the common case. I think the common case searches for shared first and then falls back to static. Would need to test pkg-config on my system to know for sure. I could test that if you could wait for ~8 hours.

iphydf commented on cmake/Findsodium.cmake:
```diff
@@ -0,0 +1,288 @@
+    )
+else()
+    if (UNIX)
+        set_target_properties(sodium PROPERTIES
```
Can you make the indentation in this file 2 spaces to match the rest of our cmake files?

iphydf replied:
Ack.

sphaerophoria commented on cmake/Dependencies.cmake:
```diff
@@ -5,19 +5,45 @@
 
-# For toxcore.
-pkg_use_module(LIBSODIUM            libsodium    )
+# for toxcore
```
fixed

sphaerophoria commented on cmake/Dependencies.cmake:
```diff
@@ -5,19 +5,45 @@
+  PKGCFG_NAME opus
+  LIB_NAMES opus
+  PATH_SUFFIXES opus
+  )
```
fixed

sphaerophoria commented on cmake/Dependencies.cmake:
```diff
@@ -5,19 +5,45 @@
+  PATH_SUFFIXES opus
+  )
+
+simple_find_package(Vpx
```
fixed

nurupo commented:
Heh, pkg-config just tells to link against `-lsodium`, both if you have only shared libsodium or only static libsodium, it doesn't specify the full filename of the library, the compiler has to figure out it on its own. Does MSVC work that way too?

```bash
$ pkg-config --libs libsodium  
-L/root/prefix/i686/lib -lsodium
```

sphaerophoria commented:
> 
> 
> Heh, pkg-config just tells to link against `-lsodium`, both if you have only shared libsodium or only static libsodium, it doesn't specify the full filename of the library, the compiler has to figure out it on its own. Does MSVC work that way too?
> 
> ```shell
> $ pkg-config --libs libsodium  
> -L/root/prefix/i686/lib -lsodium
> ```
> 
> ```shell
> $ pkg-config --static --libs libsodium  
> -L/root/prefix/i686/lib -lsodium
> ```

I'm honestly not sure. I'm way more familiar with the linux side of things than windows. I just started hacking away until things built. I read that cmake chooses the .so first with find package and falls back on static i general. Not sure what MSVC does.

I've updated the PR with shared followed by static instead of the other way around to keep compatibility with old builds. I've also changed from if (NOT sodium_FOUND) to if (NOT  TARGET sodium), i had trouble in an ubuntu 14.04 build I was trying with qtox which seems to be resolved with that solution

nurupo commented:
Welp, looks like we still managed to break the Windows cross-compilation builds: https://travis-ci.org/TokTok/c-toxcore/builds/432322743

```
Scanning dependencies of target random_testing
make[2]: *** No rule to make target 'sodium_LIBRARY_RELEASE-NOTFOUND', needed by 'random_testing.exe'.  Stop.
make[2]: *** Waiting for unfinished jobs....
make[2]: *** No rule to make target 'sodium_LIBRARY_RELEASE-NOTFOUND', needed by 'auto_file_saving_test.exe'.  Stop.
make[2]: *** Waiting for unfinished jobs....
```
